trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/witness/ExtractedLoopInvariant.java

```diff
@@ -0,0 +1,72 @@
+package de.uni_freiburg.informatik.ultimate.plugins.generator.cacsl2boogietranslator.witness;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import org.eclipse.cdt.core.dom.ast.IASTNode;
+
+import de.uni_freiburg.informatik.ultimate.boogie.ast.AssertStatement;
+import de.uni_freiburg.informatik.ultimate.boogie.ast.LoopInvariantSpecification;
+import de.uni_freiburg.informatik.ultimate.boogie.ast.Statement;
+import de.uni_freiburg.informatik.ultimate.boogie.ast.WhileStatement;
+import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.base.IDispatcher;
+import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.result.ExpressionResult;
+import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.result.ExpressionResultBuilder;
+import de.uni_freiburg.informatik.ultimate.core.lib.models.annotation.Check;
+import de.uni_freiburg.informatik.ultimate.core.model.models.ILocation;
+import de.uni_freiburg.informatik.ultimate.util.datastructures.DataStructureUtils;
+
+public class ExtractedLoopInvariant extends ExtractedWitnessInvariant {
+	public ExtractedLoopInvariant(final String invariant, final Collection<String> nodeLabel, final IASTNode match) {
+		super(invariant, nodeLabel, match);
+	}
+
+	@Override
+	public ExpressionResult transform(final ILocation loc, final IDispatcher dispatcher,
+			final ExpressionResult expressionResult) {
+		final ExpressionResult invariantExprResult = instrument(loc, dispatcher);
+		final List<Statement> statements = new ArrayList<>();
+		boolean hasLoop = false;
+		for (final Statement st : expressionResult.getStatements()) {
+			if (st instanceof WhileStatement) {
+				assert !hasLoop;
+				hasLoop = true;
+				final WhileStatement whileOld = (WhileStatement) st;
+				statements.add(new WhileStatement(loc, whileOld.getCondition(), DataStructureUtils
+						.concat(whileOld.getInvariants(), extractLoopInvariants(invariantExprResult, loc)),
+						whileOld.getBody()));
```

Comment:
Once a `while` has been instrumented, should we not  check that there is no second `while` and throw an error if there is?

Reply:
I already added an assert to check this, but throwing an exception is a better idea